Implement texture replacement support

gdkchan opened the pull request:
Overview
---

This change is implementing support for the last Patreon goal that was reached (before Patreon removed them): Texture replacement.

This allows swapping textures from a game with modified textures.

Implementation
---

Currently, textures are dumped to file on first use. When a texture is used for the first time, it computes a hash from the texture data, and tries to locate an existing file on the texture replacement folder(s). If one with a name that matches the hash (and a few other parameters) is found, it will load the file and set the texture data to the one from the file, rather than the data in memory written by the game. If none is found, and texture dumping is enabled, it will write the data into a file instead.

Currently two formats are supported: DDS and PNG.

PNG is only supported for convenience, but it is not really recommended to use as a texture format. Since it does not support 3D, Cube or Array textures, mipmaps, and it does not support compressed formats such as BC, such textures are split into several files (one per layer/slice/mipmap level), and all compressed formats are decompressed. The PNG encoder also currently favors speed over file size, which means that the saved PNGs will be larger than the ones created by other software. There is a more "normal" mode (setting `fastMode` to false) where it makes an effort to reduce file sizes, producing PNGs with sizes comparable to the ones generated by image editing software, but it will take longer to save too.

The PNG decoder currently supports indexed and non-indexed ("true color") images, of all pixel formats that PNG supports (RGB, RGBA, L, LA). It supports bit depth of 8 and lower (but I did not test bit depth lower than 8). It does not support bit depth of 16. It also does not support Adam7 interlacing, attempting to load images using either of those will print an error.

Note that all textures are exported and imported as RGBA8 when using the PNG format. The other pixel formats mentioned above are converted to RGBA8 when they are loaded.

For DDS, it supports encoding and decoding all image dimensions (2D, 2D Array, Cube, Cube Array and 3D). It supports encoding and decoding RGBA8, RGB565, RGBA5551 and RGBA4444 non-compressed formats. Currently only the compressed formats BC1-3 are supported. I plan to add support for BC4-5 soon. I'm not sure about supporting BC6-7 as not all image editing software supports those. But I plan to support at least loading BC7 compressed DDS files.

For both formats, ASTC textures are converted to RGBA8 as neither supports this compression format. Also in both cases, sRGB textures are simply reinterpreted as plain RGBA. On import, if the original texture had a sRGB format, it replaces the imported format with its sRGB variant. DDS does support sRGB formats with an extension, but not all editing software might support it properly, so it is not used right now.

Textures are loaded from multiple folders. If a texture dump path is set, it will always try to load textures from that path. In addition to that, it will also look in all mod directories that have a `textures` folder inside. If real time editing is enabled, it also sets up a `FileSystemWatcher` that monitors the texture dump folder for change on the files. If any file is changed, it will set a flag on the texture that last loaded from that file (or the texture that triggered the export that created the file), which will force its data to be re-loaded from the file the next time it is used. Note that it only ever monitors the dump folder, other texture folders are never monitored, so real time editing does not work there.

Current issues
---

Some games, such as UE4 games in general, creates various versions of the same texture with a different number of mipmap levels. This creates several duplicates when textures are dumped, which makes editing difficult since one now needs to edit several files for the same texture. We will likely need a solution to "de-duplicate" those files. My current plan is using some kind of redirection mechanism that will redirect access to lower mipmap levels to the full texture. It will likely require saving some kind of meta-data in addition to the textures themselves.

Test plan
---

I'm currently opening this PR to gather feedback from people trying the feature. From the feedback I can improve it further and fix potential issues. I think this step is important since I don't create texture packs myself, and I don't know what exactly the audience for this feature is, which tools they use, what they need from the emulator to make their life easier, etc. So hopefully I will be able to answer some of those questions with the feedback.

**Important Note:** Since this is a WIP feature, the texture dump structure is likely to change in a way that will make texture packs created with ealier versions of the PR incompatible with future versions. So it is recommended that you wait until this PR is merged before creating anything "big".

How to use
---

I wrote the usage instructions separately,
Check out the guide here: https://gist.github.com/gdkchan/33405fc557d1f29c54906bae837a42de
**Note:** The texture dumping settings are currently only implemented on the Avalonia UI, so you should use that for testing.

Testing is welcome.

Littl3Guy commented:
Changing res scale in-game isn't compatible with the dumped files it seems.
switching to 2x with dumped textures:
![2x dump](https://github.com/Ryujinx/Ryujinx/assets/115218507/72cd587b-0534-42d1-aab1-f4a77777f9a6)

without dumped textures:
![2x no dump](https://github.com/Ryujinx/Ryujinx/assets/115218507/ab9a0790-fd10-4d80-9081-963b8d1eb7a6)

Littl3Guy commented:
> Changing res scale in-game isn't compatible with the dumped files it seems. switching to 2x with dumped textures: ![2x dump](https://user-images.githubusercontent.com/115218507/264791375-72cd587b-0534-42d1-aab1-f4a77777f9a6.png)
> 
> without dumped textures: ![2x no dump](https://user-images.githubusercontent.com/115218507/264791481-ab9a0790-fd10-4d80-9081-963b8d1eb7a6.png)

Fast traveling or loading a save fixes it.

gdkchan commented:
Can you check if the issue was fixed on the latest build?

Littl3Guy commented:
> Can you check if the issue was fixed on the latest build?

Yup, that fixed the issue 👍

MelonSpeedruns commented:
It seems like Pokémon SV bushes, trees and grass blades don't move anymore with the wind with this build. Even if their textures are not replaced.

gdkchan commented:
> It seems like Pokémon SV bushes, trees and grass blades don't move anymore with the wind with this build. Even if their textures are not replaced.

Do you have a texture dump folder set? Even if you didn't actually replace any texture, it will still try to load textures from that folder. Nothing should change if there are no textures to replace with that build.

MelonSpeedruns commented:
> Do you have a texture dump folder set?

Yes I do, let me check if it happens without.

Littl3Guy commented:
> > Do you have a texture dump folder set?
> 
> Yes I do, let me check if it happens without.
> 
> EDIT: Still happens with the texture dump folder unset and deleted, and with both checkbox disabled before booting the game.

Grass animation seems to have broke between ldn 3.1.3 and master, maybe that's what you're noticing?

Littl3Guy commented:
Narrowed down the animation regression from somewhere between 1.1.813 to 1.1.865.

GamerzHell9137 commented:
Not to bring any emulator in particular but other emus when textures are being extracted are being extracted in folders of the games ID. Atm Ryujinx is extracting all stuff in the same folder which would mix with other games and its a hassle to delete the whole folder before launching another one.

gdkchan commented:
> Not to bring any emulator in particular but other emus when textures are being extracted are being extracted in folders of the games ID. Atm Ryujinx is extracting all stuff in the same folder which would mix with other games and its a hassle to delete the whole folder before launching another one.

That's because the dump folder is not meant to be a permanent directory. it is meant to be a working directory, when you're done you can copy the modified textures to the mod directory, which has one folder per game (Title ID), and you can even put as many texture replacement packs as you want there.

Foul-Tarnished commented:
A bit offtopic, but I guess we won't be able to share texture packs on the discord server ?
As it's really close to the original copyrighted material, most of the time.

Like if we used GAN models to upscale some ugly textures
All pixels are different, but it was done with the copyrighted textures, hmm

Foul-Tarnished commented:
Why are you reluctant about adding BC7 support ?
Well if you have support for loading BC7 it should be enough, probably no game use it by default.

I know that for Skyrim:SE modding, BC7 looks way better than DXT5, BC3 etc

gdkchan commented:
> Why are you reluctant about adding BC7 support ?

Because I don't know if the tools that people will use will support this format. GIMP doesn't for example. That's why I have asked for feedback. I could add it and people could start complaining the DDS is not opening in their software. That is about export of course, there's nothing preventing support for importing BC7 (but I would prefer if it exported and imported the same formats).

> Well if you have support for loading BC7 it should be enough, probably no game use it by default.

A lot of Switch games uses BC7 textures.

Foul-Tarnished commented:
So I guess if a game uses BC7, you export it as BC3 ?
So a bit of quality loss

but then the texture pack would be BC3 files replacing BC7 files, so not the same format

I don't get if the issue you're talking about is exporting in another format, or if it's loading another format than what the game use

gdkchan commented:
PR has been updated with the following changes based on feedback:
- The emulator now creates a sub-folder with the title ID of the game inside the dump folder.
  - Dump directory also no longer *needs* to exist, the emulator will create it automatically if it doesn't exist.
- Added support for exporting and importing BC4Unorm, BC5Snorm/Unorm, R8Unorm and R8G8Unorm texture formats.
  - For PNG, all of those are converted to R8G8B8A8Unorm.
  - DDS keeps the original formats.
  - BC4 and R8 are sometimes used for font textures.
  - BC5 is usually used for normal maps, without the Z component (it is calculated on the shader as `z = sqrt(1 - (x * x + y * y))`.
 - DDS now exports and imports the sRGB variants of all supported formats, instead of re-interpreting them as Unorm.
   - This means textures exported with a previous version of this PR will now appear too bright on the game, they will need to  be re-dumped.
 - DDS now exports and imports BC7 compressed textures, instead of decompressing them and exporting as R8G8B8A8Unorm.
   - Keep in mind that not all programs supports BC7. You might need to convert the texture using DirectXTex or similar tool.
 - Added support for 16-bit color channel PNGs.
   - They are converted to 8-bit on import, so there is no actual advantage in using them right now, it was added just for convenience.

blueskythlikesclouds commented:
Would it be possible to add an option to dump only ASTC textures so I can recompress them myself to avoid Ryujinx from doing it runtime?

blueskythlikesclouds commented:
> I considered adding a new format option to export using ASTC file format. It would naturally only export ASTC textures as that's the only texture format this container supports. You would need to decode and re-encode it yourself. Would that be good enough for your use case?

Yeah, that sounds great!

GamerzHell9137 commented:
Nitpick mayhaps but the default config of the emulator with the PR has texture options enabled 
![image](https://github.com/Ryujinx/Ryujinx/assets/2002038/b9f8c911-cfa5-419b-8531-f37465b3f40b)

mjuarbe23 commented:
> > Why are you reluctant about adding BC7 support ?
> 
> Because I don't know if the tools that people will use will support this format. GIMP doesn't for example. That's why I have asked for feedback. I could add it and people could start complaining the DDS is not opening in their software. That is about export of course, there's nothing preventing support for importing BC7 (but I would prefer if it exported and imported the same formats).
> 
> > Well if you have support for loading BC7 it should be enough, probably no game use it by default.
> 
> A lot of Switch games uses BC7 textures.

GIMP and Photoshop Needs a plugin for DDS BC7. i have open BC7 DDS in GIMP but it requires a Plugin. i mean its easy to find but i doubt people will due the effort to find it since its a python script

mjuarbe23 commented:
hey i have run into a issue that i need help with. so, to test the feature i dump Pokémon Brilliant Diamond to the first gym it's about 3139 files. they're all in DDS but some are in BC3 BC7 and another format. now if i save a new texture in a format that isnt the same format as the dump ryujinx wont load it in and i get Blank screen.

gdkchan commented:
> hey i have run into a issue that i need help with. so, to test the feature i dump Pokémon Brilliant Diamond to the first gym it's about 3139 files. they're all in DDS but some are in BC3 BC7 and another format. now if i save a new texture in a format that isnt the same format as the dump ryujinx wont load it in and i get Blank screen.

The emulator can read textures with formats and sizes that are different from the original, there is no such requirement. With "blank screen" do you mean that the game doesn't load at all, loads but nothing renders and you can only hear sound, or something else?

mjuarbe23 commented:
> 

so in pokemon BD when i try to load in the custom textures the intro logo where you press start its a green background with the logo and then when i go in game its just a black screen but i hear the game music playing.

gdkchan commented:
I would suggest trying to find the exact texture that causes this issue. That would help narrowing the problem down. BTW make sure that *only* the textures you want to replace are in the replacement folder. If you keep all textures there, it will replace all of them which will increase the chances of something going wrong.

mjuarbe23 commented:
> I would suggest trying to find the exact texture that causes this issue. That would help narrowing the problem down. BTW make sure that _only_ the textures you want to replace are in the replacement folder. If you keep all textures there, it will replace all of them which will increase the chances of something going wrong.

mjuarbe23 commented:
yeah i upscaled all 3139 took several days. im gonna try in a different game since someone told me of a game he tested and worked with no issues. so im gonna save it in different Formats BC3 BC7 to see if one or all gives me issues

GamerzHell9137 commented:
Games with the ModuleEngine for whatever reason do get their button prompts textures dumped.
Either that or its really well hidden in the files or maybe the files are not extracted properly/extracted at all.

Games with the engine are TOTK, Mario Wonder and Splatoon 3.

In none of the titles were the button prompts found.
UI related things in TOTK are sparsely dumped, Splatoon 3 has way too many textures to even try to find the UI related things but i've checked a lot and none of them were related to the UI. Mario Wonder has 1 texture related to button prompts but the one's that are in game are not affected by the change of it.
The texture that has the button prompts is really big but most of it is black thinking maybe the texture wasn't dumped properly?

![image](https://github.com/Ryujinx/Ryujinx/assets/2002038/705f4f3c-651c-41bd-91b5-2110abd37dfa)

Foul-Tarnished commented:
> > Would it be possible to add an option to dump only ASTC textures so I can recompress them myself to avoid Ryujinx from doing it runtime?
> 
> I considered adding a new format option to export using ASTC file format. It would naturally only export ASTC textures as that's the only texture format this container supports. You would need to decode and re-encode it yourself. Would that be good enough for your use case?

Will this get implemented relatively soon ?

Choosing between stutter and using ~10GB VRAM is frustrating

riperiperi commented:
> Games with the ModuleEngine for whatever reason do get their button prompts textures dumped. Either that or its really well hidden in the files or maybe the files are not extracted properly/extracted at all.
> 
> Games with the engine are TOTK, Mario Wonder and Splatoon 3.
> 
> In none of the titles were the button prompts found. UI related things in TOTK are sparsely dumped, Splatoon 3 has way too many textures to even try to find the UI related things but i've checked a lot and none of them were related to the UI. Mario Wonder has 1 texture related to button prompts but the one's that are in game are not affected by the change of it. The texture that has the button prompts is really big but most of it is black thinking maybe the texture wasn't dumped properly?
> 
> ![image](https://private-user-images.githubusercontent.com/2002038/277987268-705f4f3c-651c-41bd-91b5-2110abd37dfa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE4OTE3MTcsIm5iZiI6MTcwMTg5MTQxNywicGF0aCI6Ii8yMDAyMDM4LzI3Nzk4NzI2OC03MDVmNGYzYy02NTFjLTQxYmQtOTFiNS0yMTEwYWJkMzdkZmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDZUMTkzNjU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGQ4MmFjYzQxNTc4M2YzN2M1NDMwNDNlMDI5YjQ3ODE4NzAwNzMzYjMyZGE1NTVkNDg4ZGRhN2E0MjNlNDQ3NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Pp5S9-dGKYpCk8eNYMlLw2ZxoL8_yj9o9anCLPxcAus)

I think these games write a font atlas dynamically as text is drawn. If that's the case, there is no way to replace the texture, as it might be different every time. (and looking up replacements on _every_ texture modification could be very bad)

gdkchan commented:
> Can we have an option to output textures right side up?
> 
> Everything is upside down which makes textures a bit of a pain to edit in PS/GIMP.
> 
> It's not a problem for a few textures. But when you're working on a full game texture overhaul, it adds a lot of time having to transform/flip graphics.
> 
> Thanks for the consideration!

It would also need to be flipped back on import, which will take a little bit of time. And the emulator would need to store the information that the textures are flipped somewhere, to re-flip on import (not a big deal, I was already planning to include some meta data file with the texture dump for other reasons, that could be included there as well).

vbifial commented:
Hi @blueskythlikesclouds
> Would it be possible to add an option to dump only ASTC textures so I can recompress them myself to avoid Ryujinx from doing it runtime?

Mb you can use this in meantime as a temporary solution:
https://github.com/Ryujinx/Ryujinx/compare/master...vbifial:Ryujinx:texture-caching
It just dumps recompressed ASTC->BC7 textures to disk on first encountering, and then just loads them automatically even in the same session.

JeremyKingTech commented:
> > Can we have an option to output textures right side up?
> > Everything is upside down which makes textures a bit of a pain to edit in PS/GIMP.
> > It's not a problem for a few textures. But when you're working on a full game texture overhaul, it adds a lot of time having to transform/flip graphics.
> > Thanks for the consideration!
> 
> It would also need to be flipped back on import, which will take a little bit of time. And the emulator would need to store the information that the textures are flipped somewhere, to re-flip on import (not a big deal, I was already planning to include some meta data file with the texture dump for other reasons, that could be included there as well).

Thanks, it would be a huge time saver for big projects like the one I'm working on.

RabidLabMonkey commented:
![NewCanvas17](https://github.com/Ryujinx/Ryujinx/assets/77931508/6464bdfe-e11f-4be7-bb01-ddd29d69cda1)

It fails to dump the status pictures on Seven Pirates most of the time, it dumps the first one shown just fine, then it might dump the following ones (it took me like 10 retries of closing and reopening ryujinx for it to finally dump the third one, every time making a new folder so it started dumping from 0). It dumped the third texture corrupted once, that when it loaded produced the effect above. After finally getting them to dump they get replaced just fine.

Feedback:
-Adding a hotkey to trigger reloading textures would be nice, since the current setup only loads them when the original file is overwritten.

-A hotkey to turn dumping on/off during gameplay to make sorting easier.

-Separate folders for dumped and replacement textures. A lot of the time you don't want to replace every single texture but only certain ones (low resolution character textures for example) or leave out things that are not replaceable (textures generated at runtime), having everything on the same folder makes this clumsy. Also most of the time you want to keep a backup of the original textures in case you change your mind about the changes you made (or you made a mistake/the texture looks wrong when loaded).

PS. I don't think having the emulator deal with flipping images is necessary, since this is a game specific thing and just adds complexity to the thing. The only thing I can think off would need you to have the textures "right side up" is if you are doing any manual editing, since upscaling doesn't require it, and if you are manual editing textures, flipping it is 2 or 3 clicks total which adds nearly 0 extra work.

HenrikoMagnifico commented:
Artifacts seem to have expired, which makes it impossible to download compiled versions of this fork. @gdkchan would it be possible to generate new ones?

MetrosexualGarbodor commented:
@HenrikoMagnifico New artifacts are up.

HenrikoMagnifico commented:
> @HenrikoMagnifico New artifacts are up.

Thank you!

Curioso9998 commented:
I don't know if I'm doing it right, but mega man x legacy collection 1+2 doesn't start in these builds and it would be interesting to edit the textures of 3d games
![Ryujinx 1 1 0+8104714 - Mega Man X Legacy Collection 2 v1 0 0 (01005250086C4000) (64-bit) 15_01_2024 14_03_11](https://github.com/Ryujinx/Ryujinx/assets/131200760/dd8873a8-f21c-4547-8261-bd7f065d0d6f)

Kupo91 commented:
I have also had the experience that games have not started. Or that the emulator crashed when exiting a game. But I don't know whether this is due to the feature or to Avalonia.

MetrosexualGarbodor commented:
Can you please post logs of all the games you've found that don't boot on this PR, but do boot on the main build?

Curioso9998 commented:
[Ryujinx_1.1.0+8104714_2024-01-15_19-18-02.log](https://github.com/Ryujinx/Ryujinx/files/13943400/Ryujinx_1.1.0%2B8104714_2024-01-15_19-18-02.log)
[Ryujinx_1.1.0+8104714_2024-01-15_19-18-26.log](https://github.com/Ryujinx/Ryujinx/files/13943401/Ryujinx_1.1.0%2B8104714_2024-01-15_19-18-26.log)

Curioso9998 commented:
[Ryujinx_1.1.1110_2024-01-15_19-16-43.log](https://github.com/Ryujinx/Ryujinx/files/13943422/Ryujinx_1.1.1110_2024-01-15_19-16-43.log)
This is from the standard build

Kupo91 commented:
> I have also had the experience that games have not started. Or that the emulator crashed when exiting a game. But I don't know whether this is due to the feature or to Avalonia.

I tested all my games again and apart from a few which don't start on the main aswell everything started just fine. And the emulator is also not crashing anymore, if I stop emulation. So it looks like this has been fixed in the past months.

I still encountered some issues. A few of them didn't occur when I tried to record them, so maybe it was only the first start. But here is what I could record:

**Missing letters in Dragon Ball Z - Kakarot + A New Power Awakens Set:**

![ryujinx_capture_2024-01-16_17-29-38](https://github.com/Ryujinx/Ryujinx/assets/41426906/b859373a-305a-4307-bc51-e6f9cdfb2b4a)

[Ryujinx_1.1.0+8104714_2024-01-16_17-25-31.log](https://github.com/Ryujinx/Ryujinx/files/13953801/Ryujinx_1.1.0%2B8104714_2024-01-16_17-25-31.log)

**Flickering + green screen in Final Fantasy IX:**

https://github.com/Ryujinx/Ryujinx/assets/41426906/52a7e7cf-600c-4ac7-abbe-5d931858a7d5

[Ryujinx_1.1.0+8104714_2024-01-16_17-20-27.log](https://github.com/Ryujinx/Ryujinx/files/13953822/Ryujinx_1.1.0%2B8104714_2024-01-16_17-20-27.log)

**Flickering + missing letters at 1:09 in Rune Factory 5:**

https://github.com/Ryujinx/Ryujinx/assets/41426906/da4678d7-2605-4c60-8ef8-743c222e4c42

[Ryujinx_1.1.0+8104714_2024-01-16_17-34-25.log](https://github.com/Ryujinx/Ryujinx/files/13953846/Ryujinx_1.1.0%2B8104714_2024-01-16_17-34-25.log)

All these issues only occur if the option "Dump textures" is enabled. If I disable it, it works as it should.

Curioso9998 commented:
the Mega Man X Legacy Collection 1 and 2 collections do not even start
regardless of whether the texture dump function is turned on or not



Em ter., 16 de jan. de 2024 14:38, Kupo91 ***@***.***>
escreveu:

> I have also had the experience that games have not started. Or that the
> emulator crashed when exiting a game. But I don't know whether this is due
> to the feature or to Avalonia.
>
> I tested all my games again and apart from a few which don't start on the
> main aswell everything started just fine. And the emulator is also not
> crashing anymore, if I stop emulation. So it looks like this has been fixed
> in the past months.
>
> I still encountered some issues. A few of them didn't occur when I tried
> to record them, so maybe it was only the first start. But here is what I
> could record:
>
> *Missing letters in Dragon Ball Z - Kakarot + A New Power Awakens Set:*
>
> ryujinx_capture_2024-01-16_17-29-38.png (view on web)
> <https://github.com/Ryujinx/Ryujinx/assets/41426906/b859373a-305a-4307-bc51-e6f9cdfb2b4a>
>
> Ryujinx_1.1.0+8104714_2024-01-16_17-25-31.log
> <https://github.com/Ryujinx/Ryujinx/files/13953801/Ryujinx_1.1.0%2B8104714_2024-01-16_17-25-31.log>
>
> *Flickering + green screen in Final Fantasy IX:*
>
>
> https://github.com/Ryujinx/Ryujinx/assets/41426906/52a7e7cf-600c-4ac7-abbe-5d931858a7d5
>
> Ryujinx_1.1.0+8104714_2024-01-16_17-20-27.log
> <https://github.com/Ryujinx/Ryujinx/files/13953822/Ryujinx_1.1.0%2B8104714_2024-01-16_17-20-27.log>
>
> *Flickering + missing letters at 1:09 in Rune Factory 5:*
>
>
> https://github.com/Ryujinx/Ryujinx/assets/41426906/da4678d7-2605-4c60-8ef8-743c222e4c42
>
> Ryujinx_1.1.0+8104714_2024-01-16_17-34-25.log
> <https://github.com/Ryujinx/Ryujinx/files/13953846/Ryujinx_1.1.0%2B8104714_2024-01-16_17-34-25.log>
>
> All these issues only occur if the option "Dump textures" is enabled. If I
> disable it, it works as it should.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Ryujinx/Ryujinx/pull/5632#issuecomment-1894209244>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A7I7N6CIMTHNGWFRCMMR6QDYO23LDAVCNFSM6AAAAAA4GSRDASVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJUGIYDSMRUGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Curioso9998 commented:
A shortcut button to stop and start texture dumping during the game would be nice

Darkcranio95 commented:
> **Missing letters in Dragon Ball Z - Kakarot + A New Power Awakens Set:**


This only occurs on Vulkan, the letters are fine on OpenGL.

Kupo91 commented:
> > **Missing letters in Dragon Ball Z - Kakarot + A New Power Awakens Set:**
> 
> The letter problem in Kakarot occurs because Ryujinx dumps some textures with error. So when they are loaded back it doesn't appear how to supposed to. The letter one called: tex2d_3f1c5c873b0433f2529b463266c348ae_1024x1024x1_0x0 I didn't found any other solution but if you planning to texture mod Kakarot: Delete the file every time before you start the game from the dump folder. OR Dump everything you want, make your mod delete that file or dont include it in your texture pack and turn off dumping then play the game.

I don't plan to make a texture pack for this game. I just looked for bugs, so devs can make this feature better.

aarontocenter commented:
macOS -- can someone pull a .dmg build of the Texture Replacement version for macOS Apple Silicon. I would love to test this, as I use texture replacements in Dophin for so many games. I would like to clean up my own games and fix errors!

Thanks!

ekuland commented:
> macOS -- can someone pull a .dmg build of the Texture Replacement version for macOS Apple Silicon. I would love to test this, as I use texture replacements in Dophin for so many games. I would like to clean up my own games and fix errors!
> 
> Thanks!

The first reply comment always has the builds. Click the dropdown for Experimental GUI and get the macos one.

aarontocenter commented:
> > macOS -- can someone pull a .dmg build of the Texture Replacement version for macOS Apple Silicon. I would love to test this, as I use texture replacements in Dophin for so many games. I would like to clean up my own games and fix errors!
> > Thanks!
> 
> The first reply comment always has the builds. Click the dropdown for Experimental GUI and get the macos one.

How did I miss the drop down for Experimental GUI, thank you so much for pointing my attention to it. The fun begins for me tonight!

RabidLabMonkey commented:
Tested it with Unicorn Overlord (along with a few other games), texture dumping and replacement has worked well for as far I tested, only real time texture editing seems to produce issues with textures not dumping at all or dumping corrupted.

The game has no audio in this specific build, but I assume that was and issued fixed somewhere in 1.1.1X irrespective of texture replacement.

promit012345 commented:
So I am trying to download the ryujinx built for texture dumping as the normal ryujinx built do not have texture dumping. The links you guys have provided does not work. Can anyone redo it again so that I can use texture dumping for different games too on ryujinx. Thanks

Curioso9998 commented:
the texture dump can never dump all the images from the game galleries, 
![uu](https://github.com/Ryujinx/Ryujinx/assets/131200760/085764da-8fa6-482b-b1da-44ca4348ad0c)

Foul-Tarnished commented:
> Download the artifacts for this pull request:
> 
>     * [ryujinx-Release-1.1.0+8104714-linux_x64](https://nightly.link/Ryujinx/Ryujinx/actions/artifacts/1166634431.zip)
> 
>     * [ryujinx-Release-1.1.0+8104714-win_x64](https://nightly.link/Ryujinx/Ryujinx/actions/artifacts/1166641555.zip)
> 
> 
> Experimental GUI (Avalonia)
> 
>     * [ava-ryujinx-Release-1.1.0+8104714-linux_x64](https://nightly.link/Ryujinx/Ryujinx/actions/artifacts/1166634544.zip)
> 
>     * [ava-ryujinx-Release-1.1.0+8104714-macos_universal](https://nightly.link/Ryujinx/Ryujinx/actions/artifacts/1166636176.zip)
> 
>     * [ava-ryujinx-Release-1.1.0+8104714-win_x64](https://nightly.link/Ryujinx/Ryujinx/actions/artifacts/1166641678.zip)
> 
> 
> GUI-less (SDL2)
> Only for Developers

Build artifact download is expired sadly...

BLACKSugra commented:
Do you stilll have the file? looking now the checks are expired...

BLACKSugra commented:
> [Ryujinx_1.1.1110_2024-01-15_19-16-43.log](https://github.com/Ryujinx/Ryujinx/files/13943422/Ryujinx_1.1.1110_2024-01-15_19-16-43.log) 这是来自标准版本

Do you stilll have the file? looking now the checks are expired...

juza64 commented:
any chance this build will be re-uploaded? i lost it and can't access my texture project anymore:(

Foul-Tarnished commented:
> any chance this build will be re-uploaded? i lost it and can't access my texture project anymore:(

> Do you stilll have the file? looking now the checks are expired...

You'll need to compile it yourself...
```
git clone https://github.com/gdkchan/Ryujinx.git
cd Ryujinx
git checkout tex-replace
dotnet build -c Release -o build
```

And the executables will be in a folder named build, maybe need to create it first, idk

Foul-Tarnished commented:
So I dumped nearly all of TOTK,
Should I convert them to another format to have better compression/vram usage ? @gdkchan 

I imagine BC7 is great, but there are so many differents formats
(BC1_UNORM, R8G8B8A8_UNORM, BC3_UNORM_SRGB, BC5_UNORM, B5G6R5_UNORM, ...)

In which format are the astc textures saved as ? because I imagine these are uncompressed
how to handle alpha with texconv for these ?

Thanks 🙏

Foul-Tarnished commented:
Does loading custom textures has a cost ?
it seems to stutter nearly as much as loading vanilla ASTC textures

Well these are still _uncompressed_ original dumped ones

GBMasMore commented:
> > any chance this build will be re-uploaded? i lost it and can't access my texture project anymore:(
> 
> > Do you stilll have the file? looking now the checks are expired...
> 
> You'll need to compile it yourself...
> 
> ```
> git clone https://github.com/gdkchan/Ryujinx.git
> cd Ryujinx
> git checkout tex-replace
> dotnet build -c Release -o build
> ```
> 
> And the executables will be in a folder named build, maybe need to create it first, idk You might also need to install latest "Dotnet SDK"

I did tried to building it using that code, it build but whenever I press any of the 3 .exe it will crash and close intermediately. 
I double check the process of the build and did show "0 errors" at the end.

Honestly it will be better that someone re-upload the build version because even on wayback machine and any place of the internet is not available anymore.

eisnerguy1 commented:
Just checking to see if a recent build of this could be uploaded? Thanks so much 😃

ekuland commented:
> Just checking to see if a recent build of this could be uploaded? Thanks so much 😃

https://github.com/Ryujinx/Ryujinx/wiki/FAQ-and-Troubleshooting#-how-do-i-build-a-pr

GBMasMore commented:
I guess nobody will upload that build. And giving links to "build it yourself" will not help as I have mentioned months ago: It doesn't build it right.

GamerzHell9137 commented:
> I guess nobody will upload that build. And giving links to "build it yourself" will not help as I have mentioned months ago: It doesn't build it right.

Updating build is not just pressing the update button. In 6 months of development a lot of things changed which means things need to be changed here, and that takes time. The focus atm is on other things and when the devs get the time to update the build they will, until then you need to do it yourself.

If you can't build it yourself then you're doing it wrong and should probably ask around the discord or google it yourself on how to do it, its being built like every other project on github. It took me 5 minutes to download the source, open up Visual Studio which prompted me to download additional stuff to build it and the building itself which only means that you don't really care to even try to do it yourself. It seems you're only here to complain about stuff that can be done by yourself and nothing else. In the end you can stop being salty and just wait for it to be implemented, idk why you're in a hurry.

![image](https://github.com/user-attachments/assets/757284ab-45e5-4a08-a3d1-6f62c7043068)

Kupo91 commented:
> It seems you're only here to complain about stuff that can be done by yourself and nothing else. In the end you can stop being salty and just wait for it to be implemented, idk why you're in a hurry.

No, he's not the only one. I also think it's quite wrong to first advertise this feature if people just donate enough money and then not implement it if this goal is achieved.

piplup55 commented:
> No, he's not the only one. I also think it's quite wrong to first advertise this feature if people just donate enough money and then not implement it if this goal is achieved.

it's a draft PR which means it's not final, it will be implement at some point, you can't expect a feature this big to be done in a week, you have to remember the devs have life's aswell they don't just work on the emulator

GabCoolDude commented:
Could you explain what you mean by "It doesn't build right" ?

For me it builds perfectly fine. I think you might be launching the gtk build, which does not have the required ui elements to setup texture replacement, instead you should launch the avalonia build, which after you build is located in `src\Ryujinx.Ava\bin\Release\net8.0`.

Hope that helps.

Kupo91 commented:
> it's a draft PR which means it's not final, it will be implement at some point, you can't expect a feature this big to be done in a week, you have to remember the devs have life's aswell they don't just work on the emulator

It's not a week but a year. And there is no progress made since then. Me and some others made tests to help, but didn't even get a reaction on that.

Actually we know nothing about what is worked on, since there is neither a road map nor a progress report for 8 months already

gdkchan commented:
One of the reasons for this feature taking long is because I wanted it to be thoroughly tested and in the best possible shape. The main reason is that this would be the first Switch emulator with the texture replacement feature, and as such we would basically define the format for those mods, and other emulators would have to implement it. So a mistake here would either cause problems for other emulators, or motivate them to implement a different format that is incompatible with this emulator. Anyway, I admit this got sidetracked, and I'll try to move it forward.

There are a few features that I'd like to remove, because from feedback it doesn't look like anyone is using/expecting them, and removing it would make the implementation simpler:
- Real-time texture editing.
  - The logic to detect modifications is a bit tricky to avoid the emulator stepping on its own toes (it should not detect a texture it just wrote as modified etc).
- Texture replacement inside mods folder.
  - People does not expect it to work this way, and it is not a "standard" mod structure (it will never work on actual Switch hardware).
- Maybe separate the folders used for dumping and loading textures?

If someone is making use of one of those features and feels that is important to keep them, please let me know.

Things that I plan to add:
- ASTC texture format support. Someone requested it.

I'll take a look at the issues that were reported and see which ones are actionable.

BTW, the PR has been rebased, but I did not test it thoroughly yet.

Kupo91 commented:
> There are a few features that I'd like to remove, because from feedback it doesn't look like anyone is using/expecting them, and removing it would make the implementation simpler:
> 
>     * Real-time texture editing.
>       
>       * The logic to detect modifications is a bit tricky to avoid the emulator stepping on its own toes (it should not detect a texture it just wrote as modified etc).
> 
>     * Texture replacement inside mods folder.
>       
>       * People does not expect it to work this way, and it is not a "standard" mod structure (it will never work on actual Switch hardware).
> 
>     * Maybe separate the folders used for dumping and loading textures?

Thanks for the explanation.

I already worked with the texture replacement features of other emulators and would welcome the changes, you mentioned.
If you add separated folders for texture dumps and texture replacements, it would be nice if the emulator could check what is inside the replacements folder and skip these texture dumps. 
And no need for real time texture editing but a hotkey to reload texture replacements would be useful.

juza64 commented:
1: perhaps like Kupo91 mentioned a hotkey to reload textures may be a good idea. Ryujinx doesn’t have save states, so relaunching a game and returning to the texture being worked on can be time consuming. 

2:I’m indifferent to folder location. 

3: separate folders would be very nice. Perhaps even searching sub folders for textures to keep better organization if possible. N64 and ps2 emu I use have separate folders and it helps keeping everything  neat.
